plugins/ltac2/tac2ffi.mli

```diff
@@ -24,66 +24,122 @@ val make_repr : ('a -> valexpr) -> (valexpr -> 'a) -> 'a repr
 
 val map_repr : ('a -> 'b) -> ('b -> 'a) -> 'a repr -> 'b repr
 
+(** Type annotations. *)
+
+module Types : sig
+
+  type t
+  (** Type of an argument of an external. This is basically
+      [int Tac2expr.glb_typexpr] but we may add some smart constructors
+      so that the user doesn't need to pick a number for variables used only once. *)
+
+  val any : t
+
+  val default : t option -> t
+
+  val var : int -> t
+
+  val tuple : t list -> t
+
+  val (!) : Tac2expr.type_constant -> t list -> t
+
+  val (!!) : Tac2expr.type_constant -> t
+  (** Alias of [(!)] for 0-ary type formers. *)
+
+  val (@->) : t -> t -> t
+
+  val as_scheme : t -> Tac2expr.type_scheme
+
+end
+
+type 'a annotated = 'a * Types.t option
+
+val typed : 'r -> Types.t -> 'r annotated
+val untyped : 'r -> 'r annotated
+
 (** These functions allow to convert back and forth between OCaml and Ltac2
     data representation. The [to_*] functions raise an anomaly whenever the data
-    has not expected shape. *)
+    has not expected shape.
+
+    Unless no default annotation is useful, [foo_] is the raw converter
+    and [foo] annotates it with types taking any necessary type arguments.
+*)
 
 val of_unit : unit -> valexpr
 val to_unit : valexpr -> unit
-val unit : unit repr
+val unit : unit repr annotated
+val unit_ : unit repr
```

Comment:
I wrote everything following this pattern (`foo_` is the raw repr, `foo` is the annotated repr unless no annotation works such as `valexpr` in which case `foo` is the raw repr).
I thought the raw reprs might be used in further constructions but this turned out to be rare. There's 1 case where we use `bool` repr but the expected type is `Std.orientation` (`tac_setoid_rewrite`) and that could easily have been written in terms of replacing the type of the annotated repr. `to_repr` / `of_repr` could also take annotated repr and anyway are also not very much used directly.